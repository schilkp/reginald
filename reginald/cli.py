from collections.abc import Callable
from dataclasses import dataclass
from typing import List

from typing_extensions import Self

from reginald.datamodel import RegisterMap


@dataclass
class CLI:
    input_file: str
    output_path: str
    generator: Callable[[RegisterMap, Self], None]
    generator_args: List[str]
