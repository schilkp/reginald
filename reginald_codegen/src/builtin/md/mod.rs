use crate::{error::Error, utils::table_col_width};
use std::fmt::Write;

pub mod datasheet;

pub fn md_table(out: &mut dyn Write, rows: &Vec<Vec<String>>) -> Result<(), Error> {
    if rows.len() < 2 {
        return Err(Error::GeneratorError(format!(
            "Cannot generate markdown table from {} rows, need at least 2.",
            rows.len()
        )));
    }

    let col_widths = table_col_width(rows);

    md_table_row(out, &rows[0], &col_widths)?;
    md_sep_row(out, &col_widths)?;
    for col in rows.iter().skip(1) {
        md_table_row(out, col, &col_widths)?;
    }

    Ok(())
}

fn md_table_row(out: &mut dyn Write, row: &[String], widths: &[usize]) -> Result<(), Error> {
    write!(out, "| ")?;
    for (col_idx, width) in widths.iter().enumerate() {
        let col = row.get(col_idx).map_or(String::new(), std::borrow::ToOwned::to_owned);
        if col_idx != 0 {
            write!(out, " | ")?;
        }
        write!(out, "{col: <width$}")?;
    }
    writeln!(out, " |")?;

    Ok(())
}

fn md_sep_row(out: &mut dyn Write, widths: &[usize]) -> Result<(), Error> {
    write!(out, "| ")?;
    for (col_idx, width) in widths.iter().enumerate() {
        if col_idx != 0 {
            write!(out, " | ")?;
        }
        for _ in 0..*width {
            write!(out, "-")?;
        }
    }
    writeln!(out, " |")?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_md_table() {
        let mut is = String::new();

        md_table(
            &mut is,
            &vec![
                vec!["1".into(), "22222".into(), "3".into()],
                vec!["A".into(), "B".into(), "CCCC".into()],
                vec![":)".into(), "!".into()],
            ],
        )
        .unwrap();
        let should = "| 1  | 22222 | 3    |\n\
                      | -- | ----- | ---- |\n\
                      | A  | B     | CCCC |\n\
                      | :) | !     |      |\n";
        assert_eq!(is, should);
    }
}
