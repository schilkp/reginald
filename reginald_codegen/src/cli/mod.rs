mod cmd;
mod diff;

use std::process::ExitCode;

use clap::Parser;

#[derive(Parser, Debug)]
#[command(version, about)]
#[command(name = "reginald")]
enum Cli {
    Gen(cmd::gen::Command),
    Completion(cmd::completion::Command),
    Tool(cmd::tool::Command),
}

pub fn cli_main() -> ExitCode {
    let cli = Cli::parse();

    let err = match cli {
        Cli::Gen(gen) => cmd::gen::cmd(gen),
        Cli::Completion(c) => cmd::completion::cmd(c),
        Cli::Tool(tool) => cmd::tool::cmd(tool),
    };

    match err {
        Ok(_) => ExitCode::SUCCESS,
        Err(err) => {
            eprintln!("{}", err);
            ExitCode::FAILURE
        }
    }
}
