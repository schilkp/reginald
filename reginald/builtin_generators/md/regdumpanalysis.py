import os
from typing import Dict, List, Union

import yaml
from pydantic import ValidationError
from pydantic.dataclasses import dataclass
from tabulate import tabulate
from yaml import SafeLoader

from reginald.cli import CLI
from reginald.datamodel import *
from reginald.error import ReginaldException
from reginald.generator import OutputGenerator


@dataclass
class YamlBinaryDump:
    binary: Dict[NonNegativeInt, Union[List[NonNegativeInt], NonNegativeInt]]

    @classmethod
    def from_yaml_file(cls, file_name: str):
        try:
            with open(file_name) as f:
                data = yaml.load(f, Loader=SafeLoader)
                return YamlBinaryDump(**data)
        except FileNotFoundError:
            raise ReginaldException(f"File {file_name} not found")
        except ValidationError as e:
            raise ReginaldException(str(e))

    def flatten(self) -> Dict[NonNegativeInt, NonNegativeInt]:
        dump = {}  # type: Dict[NonNegativeInt, NonNegativeInt]

        for at, values in self.binary.items():

            if isinstance(values, List):
                for value in values:
                    if at in dump:
                        raise ReginaldException(f"YamlBinaryDump has two values at address {at}!")
                    dump[at] = value
                    at = at+1

            else:
                if at in dump:
                    raise ReginaldException(f"YamlBinaryDump has two values at address {at}!")
                dump[at] = values

        return dump


class Generator(OutputGenerator):
    @classmethod
    def description(cls):
        return "TODO"

    @classmethod
    def generate(cls, rmap: RegisterMap, cli: CLI):

        if len(cli.generator_args) != 1:
            raise ReginaldException("md_dumpanalysis requires a yaml binary dump as it's only argument")

        dump_yaml = YamlBinaryDump.from_yaml_file(cli.generator_args[0])
        dump = dump_yaml.flatten()
        adrs = sorted(dump.keys())

        out = []

        # Generate header:
        out.append(f"# {rmap.map_name} Register Dump Analysis")
        out.append(f"")
        out.append(f"")

        for adr in adrs:
            register_name = rmap.get_registername_at(adr)

            if register_name is not None:
                reg = rmap.registers[register_name]
                out.append(f"## 0x{adr:0X} - {register_name}")
                out.append(f"  - 0x{dump[adr]:X}")
                out.append(f"  - 0b{dump[adr]:b}")

                # Collect all bitranges that make up this register - field or not:
                register_bitranges = []  # type: List[BitRange]
                for field in reg.fields.values():
                    for range in field.get_bitranges():
                        register_bitranges.append(range)
                register_bitranges.extend(reg.get_unused_bits(include_always_write=True).get_bitranges())

                # Sort bitranges:
                register_bitranges = sorted(register_bitranges, key=lambda x: x.lsb_position, reverse=True)

                bitrow = ["Bits:"]
                field_row = ["Field:"]
                value_row = ["Value:"]
                decode_row = ["Decode:"]

                for bitrange in register_bitranges:
                    bitrow.append(str(bitrange))

                    field_val = bitrange.extract_this_field_from(dump[adr])

                    value_row.append(f"0x{field_val:X}")

                    # Retrieve field that coresponds to this range (if any):
                    field_name = reg.get_fieldname_at(bitrange.lsb_position)

                    if field_name is not None:
                        field_row.append(field_name)

                        # Lookup if this value in this value coresponds to an enum:
                        field = reg.fields[field_name]
                        if field.enum is not None:
                            enum_entryname = field.lookup_enum_entry_name(field_val)
                            if enum_entryname is not None:
                                decode_row.append(f"{enum_entryname} (0x{field_val:X})")
                            else:
                                decode_row.append(f"DECODE ERROR")
                        else:
                            decode_row.append(f"?")

                    elif reg.is_bit_always_write(bitrange.lsb_position):
                        val = reg.get_always_write_value(Bits.from_bitrange(bitrange))
                        field = f"Always write 0x{val:x}"
                        field_row.append(field)
                        if val == field_val:
                            decode_row.append(f"OK")
                        else:
                            decode_row.append(f"ERROR")
                    else:
                        field_row.append("?")
                        decode_row.append(f"?")

                out.append("")
                out.append(tabulate([bitrow, field_row, value_row, decode_row], headers="firstrow",
                                    tablefmt="pipe", numalign="center", stralign="center"))
                out.append("")

                # Field info:
                out.append(f"*Bitfields*:")

                for field_name, field in reg.fields.items():
                    field_val = field.bits.extract_this_field_from(dump[adr])
                    out.append(f"   - {field_name}: 0x{field_val:X}")
                    out.extend(field.docs.two_line(prefix="  -"))

                    if field.enum is not None:
                        enum_entryname = field.lookup_enum_entry_name(field_val)
                        if enum_entryname is not None:
                            entry = field.enum.entries[enum_entryname]
                            out.append(f"       - *SELECTED*: {enum_entryname}")
                            out.extend(entry.docs.two_line(prefix="         - "))
                        else:
                            decode_row.append(
                                f"       - *DECODE ERROR*: This field accepts an enum, but it's value does not correspond to any enum entry.")
                    else:
                        decode_row.append(f"?")

            else:
                out.append(f"## 0x{adr:0X} - ?")
                out.append(f"   - 0x{dump[adr]:X}")
                out.append(f"   - 0b{dump[adr]:b}")

            out.append(f"")
            out.append(f"")

        output_file = os.path.join(cli.output_path, f"{rmap.map_name.lower()}_analysis.md")
        with open(output_file, 'w') as outfile:
            outfile.write("\n".join(out))
        print(f"Generated {output_file}...")
