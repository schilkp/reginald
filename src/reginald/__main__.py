import sys

from reginald.cli import parse_args
from reginald.error import ReginaldException
from reginald.input.convert_yaml import YAMLConverter
from reginald.input.parse_yaml import YAML_RegisterMap
from reginald.input.validate_map import MapValidator


def main():

    try:

        # Parse command line args:
        cli, generator = parse_args()

        # Open, parse, and validate input file:
        r = YAML_RegisterMap.from_yaml_file(cli.input_file)
        r = YAMLConverter(r).convert()
        MapValidator(r).validate()

        # Generate output using selected generator:
        generator.generate(r, cli.input_file, cli.output_file, cli.generator_args)

    except ReginaldException as e:
        print(e, file=sys.stderr)
        exit(-1)

    exit(0)


if __name__ == '__main__':
    main()
