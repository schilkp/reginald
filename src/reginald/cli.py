import argparse
from dataclasses import dataclass
from typing import List, Tuple

import reginald.builtin_generators.c.funcpack
import reginald.builtin_generators.c.macromap
import reginald.builtin_generators.external_jinja2_template
import reginald.builtin_generators.md.doc
import reginald.builtin_generators.md.regdumpanalysis
from reginald.error import ReginaldException
from reginald.generator import OutputGenerator

builtin_generators = {
    'c.macromap': reginald.builtin_generators.c.macromap.Generator(),
    'c.funcpack': reginald.builtin_generators.c.funcpack.Generator(),
    'md.regdumpanalysis': reginald.builtin_generators.md.regdumpanalysis.Generator(),
    'md.doc': reginald.builtin_generators.md.doc.Generator(),
    'jinja2': reginald.builtin_generators.external_jinja2_template.Generator()
}


@dataclass
class CLI:
    input_file: str
    output_file: str
    generator_args: List[str]


def parse_args() -> Tuple[CLI, OutputGenerator]:

    builtin_choices_text = []
    for name, generator in builtin_generators.items():
        builtin_choices_text.append(f"   {name}: {generator.description()}")

    builtin_choices_text = "\n".join(builtin_choices_text)

    parser = argparse.ArgumentParser(prog="Reginald",
                                     description='Register map utility.\nPhilipp Schilk, 2022-2023',
                                     epilog="builtin generators: \n" + builtin_choices_text,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument('--version', action='version', version='reginald ' + reginald.__version__)
    parser.add_argument('input_file',
                        help="input register description yaml")
    parser.add_argument('output_file',
                        help=f"name of file to be generated")
    parser.add_argument('output_generator',
                        help=f"builtin generator to use")
    parser.add_argument('generator_args', nargs=argparse.REMAINDER,
                        help="additional arguments passed to the selected output generator")

    args = parser.parse_args()

    selection = args.output_generator
    if selection in builtin_generators:
        generator = builtin_generators[selection]
    else:
        raise ReginaldException("Error: Unknown generator.")

    return CLI(input_file=args.input_file,
               output_file=args.output_file,
               generator_args=args.generator_args), generator
