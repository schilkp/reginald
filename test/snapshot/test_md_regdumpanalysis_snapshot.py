import os

import reginald.builtin_generators.md.regdumpanalysis as builtin_md_regdumpanalysis
from reginald.__main__ import generate
from reginald.cli import CLI

SNAPSHOT_DIR = os.path.dirname(__file__)


def test_md_regdumpanalysis_8b(snapshot):
    generator = builtin_md_regdumpanalysis.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_8b.yaml"),
              output_file="md_regdumpanalysis_8b.md",
              generator_args=[os.path.join(SNAPSHOT_DIR, "md_regdumpanalysis_8b.yaml")],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_regdumpanalysis_9b(snapshot):
    generator = builtin_md_regdumpanalysis.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_9b.yaml"),
              output_file="md_regdumpanalysis_9b.md",
              generator_args=[os.path.join(SNAPSHOT_DIR, "md_regdumpanalysis_9b.yaml")],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_regdumpanalysis_32b(snapshot):
    generator = builtin_md_regdumpanalysis.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_32b.yaml"),
              output_file="md_regdumpanalysis_32b.md",
              generator_args=[os.path.join(SNAPSHOT_DIR, "md_regdumpanalysis_32b.yaml")],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot


def test_md_regdumpanalysis_64b(snapshot):
    generator = builtin_md_regdumpanalysis.Generator()
    cli = CLI(input_file=os.path.join(SNAPSHOT_DIR, "input_64b.yaml"),
              output_file="md_regdumpanalysis_64b.md",
              generator_args=[os.path.join(SNAPSHOT_DIR, "md_regdumpanalysis_64b.yaml")],
              verify=False)
    output = "\n".join(generate(cli, generator))
    assert output == snapshot
