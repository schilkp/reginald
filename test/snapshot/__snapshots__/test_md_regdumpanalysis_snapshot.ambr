# serializer version: 1
# name: test_md_regdumpanalysis_32b
  '''
  # Snapshotter32B Register Dump Analysis
  
  
  ## 0x0 - BASIC
    - 0xFF000100
    - 0b11111111000000000000000100000000
  
  |  Bits:  |   31-9   |  8  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:--------:|:---:|:------:|:------:|:------:|:------:|:------:|
  | Field:  |  FIELD0  |  ?  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Value:  | 0x7F8000 | 0x1 |  0x0   |  0x0   |  0x0   |  0x0   |  0x0   |
  | Decode: |    ?     |  ?  |   ?    |   ?    |   ?    |   ?    |   ?    |
  
  *Bitfields*:
     - FIELD0: 0x7F8000
     - FIELD1: 0x0
     - FIELD2: 0x0
     - FIELD3: 0x0
     - FIELD4: 0x0
     - FIELD5: 0x0
  
  
  ## 0x1 - ?
     - 0xFF000101
     - 0b11111111000000000000000100000001
  
  
  ## 0xF - COMPLICATED
    - 0x1AB
    - 0b110101011
  
  |  Bits:  |  31-8  |      7-6       |       5-4        |  3-0   |
  |:-------:|:------:|:--------------:|:----------------:|:------:|
  | Field:  |   ?    |     FIELD1     | Always write 0x1 | FIELD2 |
  | Value:  |  0x1   |      0x2       |       0x2        |  0xB   |
  | Decode: |   ?    | NOT_COOL (0x2) |      ERROR       | ERROR  |
  
  *Bitfields*:
     - FIELD1: 0x2
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
     - FIELD2: 0xB
       - Another cool field.
       - Important details.
  
  
  ## 0x10 - CH1_CNFG
    - 0xEF
    - 0b11101111
  
  |  Bits:  |  31-30  |  29-1  |    0     |
  |:-------:|:-------:|:------:|:--------:|
  | Field:  |    Y    |   ?    |    A     |
  | Value:  |   0x0   |  0x77  |   0x1    |
  | Decode: |    ?    |   ?    | ON (0x1) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x11 - CH1_STATUS
    - 0x12
    - 0b10010
  
  |  Bits:  |  31-3  |      2-0       |
  |:-------:|:------:|:--------------:|
  | Field:  |   ?    |     STATUS     |
  | Value:  |  0x2   |      0x2       |
  | Decode: |   ?    | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  ## 0x20 - CH2_CNFG
    - 0x143
    - 0b101000011
  
  |  Bits:  |  31-30  |  29-1  |    0     |
  |:-------:|:-------:|:------:|:--------:|
  | Field:  |    Y    |   ?    |    A     |
  | Value:  |   0x0   |  0xA1  |   0x1    |
  | Decode: |    ?    |   ?    | ON (0x1) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x21 - CH2_STATUS
    - 0x150
    - 0b101010000
  
  |  Bits:  |  31-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Value:  |  0x2A  |  0x0   |
  | Decode: |   ?    | ERROR  |
  
  *Bitfields*:
     - STATUS: 0x0
       - Status
  
  
  ## 0x30 - CH3_CNFG
    - 0x90
    - 0b10010000
  
  |  Bits:  |  31-30  |  29-1  |     0     |
  |:-------:|:-------:|:------:|:---------:|
  | Field:  |    Y    |   ?    |     A     |
  | Value:  |   0x0   |  0x48  |    0x0    |
  | Decode: |    ?    |   ?    | OFF (0x0) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x0
       - Controls value A
         - *SELECTED*: OFF
  
  
  ## 0x31 - CH3_STATUS
    - 0x112
    - 0b100010010
  
  |  Bits:  |  31-3  |      2-0       |
  |:-------:|:------:|:--------------:|
  | Field:  |   ?    |     STATUS     |
  | Value:  |  0x22  |      0x2       |
  | Decode: |   ?    | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  '''
# ---
# name: test_md_regdumpanalysis_64b
  '''
  # Snapshotter64B Register Dump Analysis
  
  
  ## 0x0 - BASIC
    - 0xFF00FF00FF000100
    - 0b1111111100000000111111110000000011111111000000000000000100000000
  
  |  Bits:  |       63-9       |  8  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:----------------:|:---:|:------:|:------:|:------:|:------:|:------:|
  | Field:  |      FIELD0      |  ?  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Value:  | 0x7F807F807F8000 | 0x1 |  0x0   |  0x0   |  0x0   |  0x0   |  0x0   |
  | Decode: |        ?         |  ?  |   ?    |   ?    |   ?    |   ?    |   ?    |
  
  *Bitfields*:
     - FIELD0: 0x7F807F807F8000
     - FIELD1: 0x0
     - FIELD2: 0x0
     - FIELD3: 0x0
     - FIELD4: 0x0
     - FIELD5: 0x0
  
  
  ## 0x1 - ?
     - 0xFF00FF00FF000101
     - 0b1111111100000000111111110000000011111111000000000000000100000001
  
  
  ## 0xF - COMPLICATED
    - 0x1AB
    - 0b110101011
  
  |  Bits:  |  63-8  |      7-6       |       5-4        |  3-0   |
  |:-------:|:------:|:--------------:|:----------------:|:------:|
  | Field:  |   ?    |     FIELD1     | Always write 0x1 | FIELD2 |
  | Value:  |  0x1   |      0x2       |       0x2        |  0xB   |
  | Decode: |   ?    | NOT_COOL (0x2) |      ERROR       | ERROR  |
  
  *Bitfields*:
     - FIELD1: 0x2
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
     - FIELD2: 0xB
       - Another cool field.
       - Important details.
  
  
  ## 0x10 - CH1_CNFG
    - 0xEF
    - 0b11101111
  
  |  Bits:  |  63-30  |  29-1  |    0     |
  |:-------:|:-------:|:------:|:--------:|
  | Field:  |    Y    |   ?    |    A     |
  | Value:  |   0x0   |  0x77  |   0x1    |
  | Decode: |    ?    |   ?    | ON (0x1) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x11 - CH1_STATUS
    - 0x12
    - 0b10010
  
  |  Bits:  |  63-3  |      2-0       |
  |:-------:|:------:|:--------------:|
  | Field:  |   ?    |     STATUS     |
  | Value:  |  0x2   |      0x2       |
  | Decode: |   ?    | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  ## 0x20 - CH2_CNFG
    - 0x143
    - 0b101000011
  
  |  Bits:  |  63-30  |  29-1  |    0     |
  |:-------:|:-------:|:------:|:--------:|
  | Field:  |    Y    |   ?    |    A     |
  | Value:  |   0x0   |  0xA1  |   0x1    |
  | Decode: |    ?    |   ?    | ON (0x1) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x21 - CH2_STATUS
    - 0x150
    - 0b101010000
  
  |  Bits:  |  63-3  |  2-0   |
  |:-------:|:------:|:------:|
  | Field:  |   ?    | STATUS |
  | Value:  |  0x2A  |  0x0   |
  | Decode: |   ?    | ERROR  |
  
  *Bitfields*:
     - STATUS: 0x0
       - Status
  
  
  ## 0x30 - CH3_CNFG
    - 0x90
    - 0b10010000
  
  |  Bits:  |  63-30  |  29-1  |     0     |
  |:-------:|:-------:|:------:|:---------:|
  | Field:  |    Y    |   ?    |     A     |
  | Value:  |   0x0   |  0x48  |    0x0    |
  | Decode: |    ?    |   ?    | OFF (0x0) |
  
  *Bitfields*:
     - Y: 0x0
     - A: 0x0
       - Controls value A
         - *SELECTED*: OFF
  
  
  ## 0x31 - CH3_STATUS
    - 0x112
    - 0b100010010
  
  |  Bits:  |  63-3  |      2-0       |
  |:-------:|:------:|:--------------:|
  | Field:  |   ?    |     STATUS     |
  | Value:  |  0x22  |      0x2       |
  | Decode: |   ?    | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  '''
# ---
# name: test_md_regdumpanalysis_8b
  '''
  # Snapshotter8B Register Dump Analysis
  
  
  ## 0x0 - BASIC
    - 0x0
    - 0b0
  
  |  Bits:  |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Value:  |  0x0   |  0x0   |  0x0   |  0x0   |  0x0   |
  | Decode: |   ?    |   ?    |   ?    |   ?    |   ?    |
  
  *Bitfields*:
     - FIELD1: 0x0
     - FIELD2: 0x0
     - FIELD3: 0x0
     - FIELD4: 0x0
     - FIELD5: 0x0
  
  
  ## 0x1 - ?
     - 0x1
     - 0b1
  
  
  ## 0xF - COMPLICATED
    - 0xAB
    - 0b10101011
  
  |  Bits:  |      7-6       |       5-4        |  3-0   |
  |:-------:|:--------------:|:----------------:|:------:|
  | Field:  |     FIELD1     | Always write 0x1 | FIELD2 |
  | Value:  |      0x2       |       0x2        |  0xB   |
  | Decode: | NOT_COOL (0x2) |      ERROR       | ERROR  |
  
  *Bitfields*:
     - FIELD1: 0x2
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
     - FIELD2: 0xB
       - Another cool field.
       - Important details.
  
  
  ## 0x10 - CH1_CNFG
    - 0xEF
    - 0b11101111
  
  |  Bits:  |  7-1  |    0     |
  |:-------:|:-----:|:--------:|
  | Field:  |   ?   |    A     |
  | Value:  | 0x77  |   0x1    |
  | Decode: |   ?   | ON (0x1) |
  
  *Bitfields*:
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x11 - CH1_STATUS
    - 0x12
    - 0b10010
  
  |  Bits:  |  7-3  |      2-0       |
  |:-------:|:-----:|:--------------:|
  | Field:  |   ?   |     STATUS     |
  | Value:  |  0x2  |      0x2       |
  | Decode: |   ?   | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  ## 0x20 - CH2_CNFG
    - 0x43
    - 0b1000011
  
  |  Bits:  |  7-1  |    0     |
  |:-------:|:-----:|:--------:|
  | Field:  |   ?   |    A     |
  | Value:  | 0x21  |   0x1    |
  | Decode: |   ?   | ON (0x1) |
  
  *Bitfields*:
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x21 - CH2_STATUS
    - 0x50
    - 0b1010000
  
  |  Bits:  |  7-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Value:  |  0xA  |  0x0   |
  | Decode: |   ?   | ERROR  |
  
  *Bitfields*:
     - STATUS: 0x0
       - Status
  
  
  ## 0x30 - CH3_CNFG
    - 0x90
    - 0b10010000
  
  |  Bits:  |  7-1  |     0     |
  |:-------:|:-----:|:---------:|
  | Field:  |   ?   |     A     |
  | Value:  | 0x48  |    0x0    |
  | Decode: |   ?   | OFF (0x0) |
  
  *Bitfields*:
     - A: 0x0
       - Controls value A
         - *SELECTED*: OFF
  
  
  ## 0x31 - CH3_STATUS
    - 0x12
    - 0b10010
  
  |  Bits:  |  7-3  |      2-0       |
  |:-------:|:-----:|:--------------:|
  | Field:  |   ?   |     STATUS     |
  | Value:  |  0x2  |      0x2       |
  | Decode: |   ?   | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  '''
# ---
# name: test_md_regdumpanalysis_9b
  '''
  # Snapshotter9B Register Dump Analysis
  
  
  ## 0x0 - BASIC
    - 0x100
    - 0b100000000
  
  |  Bits:  |   8    |  7-6   |   5    |  4-3   |  2-1   |   0    |
  |:-------:|:------:|:------:|:------:|:------:|:------:|:------:|
  | Field:  | FIELD0 | FIELD1 | FIELD2 | FIELD3 | FIELD4 | FIELD5 |
  | Value:  |  0x1   |  0x0   |  0x0   |  0x0   |  0x0   |  0x0   |
  | Decode: |   ?    |   ?    |   ?    |   ?    |   ?    |   ?    |
  
  *Bitfields*:
     - FIELD0: 0x1
     - FIELD1: 0x0
     - FIELD2: 0x0
     - FIELD3: 0x0
     - FIELD4: 0x0
     - FIELD5: 0x0
  
  
  ## 0x1 - ?
     - 0x101
     - 0b100000001
  
  
  ## 0xF - COMPLICATED
    - 0x1AB
    - 0b110101011
  
  |  Bits:  |  8-6   |       5-4        |  3-0   |
  |:-------:|:------:|:----------------:|:------:|
  | Field:  | FIELD1 | Always write 0x1 | FIELD2 |
  | Value:  |  0x6   |       0x2        |  0xB   |
  | Decode: | ERROR  |      ERROR       | ERROR  |
  
  *Bitfields*:
     - FIELD1: 0x6
     - FIELD2: 0xB
       - Another cool field.
       - Important details.
  
  
  ## 0x10 - CH1_CNFG
    - 0xEF
    - 0b11101111
  
  |  Bits:  |   8-7    |  6-0  |
  |:-------:|:--------:|:-----:|
  | Field:  |    A     |   ?   |
  | Value:  |   0x1    | 0x6F  |
  | Decode: | ON (0x1) |   ?   |
  
  *Bitfields*:
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x11 - CH1_STATUS
    - 0x12
    - 0b10010
  
  |  Bits:  |  8-3  |      2-0       |
  |:-------:|:-----:|:--------------:|
  | Field:  |   ?   |     STATUS     |
  | Value:  |  0x2  |      0x2       |
  | Decode: |   ?   | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  ## 0x20 - CH2_CNFG
    - 0x143
    - 0b101000011
  
  |  Bits:  |  8-7  |  6-0  |
  |:-------:|:-----:|:-----:|
  | Field:  |   A   |   ?   |
  | Value:  |  0x2  | 0x43  |
  | Decode: | ERROR |   ?   |
  
  *Bitfields*:
     - A: 0x2
       - Controls value A
  
  
  ## 0x21 - CH2_STATUS
    - 0x150
    - 0b101010000
  
  |  Bits:  |  8-3  |  2-0   |
  |:-------:|:-----:|:------:|
  | Field:  |   ?   | STATUS |
  | Value:  | 0x2A  |  0x0   |
  | Decode: |   ?   | ERROR  |
  
  *Bitfields*:
     - STATUS: 0x0
       - Status
  
  
  ## 0x30 - CH3_CNFG
    - 0x90
    - 0b10010000
  
  |  Bits:  |   8-7    |  6-0  |
  |:-------:|:--------:|:-----:|
  | Field:  |    A     |   ?   |
  | Value:  |   0x1    | 0x10  |
  | Decode: | ON (0x1) |   ?   |
  
  *Bitfields*:
     - A: 0x1
       - Controls value A
         - *SELECTED*: ON
  
  
  ## 0x31 - CH3_STATUS
    - 0x112
    - 0b100010010
  
  |  Bits:  |  8-3  |      2-0       |
  |:-------:|:-----:|:--------------:|
  | Field:  |   ?   |     STATUS     |
  | Value:  | 0x22  |      0x2       |
  | Decode: |   ?   | NOT_COOL (0x2) |
  
  *Bitfields*:
     - STATUS: 0x2
       - Status
         - *SELECTED*: NOT_COOL
           - All not so cool.
           - Should be worried.
  
  
  '''
# ---
